SF Salaries Exercise

Welcome to a quick exercise for you to practice your pandas skills! We will be using the SF Salaries Dataset from Kaggle! Just follow along and complete the tasks outlined in bold below. The tasks will get harder and harder as you go along.

1. Read Salaries.csv as a dataframe called sal.

In [3]:
# Import the pandas library with the alias 'pd'.
import pandas as pd

# Read the data from the CSV file "Salaries.csv" into a DataFrame 'sal'.
sal = pd.read_csv("Salaries.csv")

2.Check the head of the DataFrame.

In [4]:
# Display the first few rows of the DataFrame 'sal' using the 'head()' method.
sal.head()

Id       EmployeeName                                        JobTitle  \
0   1     NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2       GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3     ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3   4  CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4   5    PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   

     BasePay  OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  \
0  167411.18         0.00  400184.25       NaN  567595.43         567595.43   
1  155966.02    245131.88  137811.38       NaN  538909.28         538909.28   
2  212739.13    106088.18   16452.60       NaN  335279.91         335279.91   
3   77916.00     56120.71  198306.90       NaN  332343.61         332343.61   
4  134401.60      9737.00  182234.59       NaN  326373.19         326373.19   

   Year  Notes         Agency  Status  
0  2011    NaN  San Francisco     NaN  
1  2011    NaN  San Francisco     NaN  
2  2011    NaN  San Francisco     NaN  
3  2011    NaN  San Francisco     NaN  
4  2011    NaN  San Francisco     NaN

3.Use the .info() method to find out how many entries there are

In [5]:
sal.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 13 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  object 
 2   JobTitle          148654 non-null  object 
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          112491 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Notes             0 non-null       float64
 11  Agency            148654 non-null  object 
 12  Status            0 non-null       float64
dtypes: float64(8), int64(2), object(3)
memory usage: 14.7+ MB


4.What is the average BasePay ?
a. 66325.44884050643

In [6]:
# Calculate the mean (average) of the 'BasePay' column in the DataFrame 'sal'.
average_base_pay = sal['BasePay'].mean()
print(average_base_pay)

66325.4488404877


5.What is the highest amount of OvertimePay in the dataset ?
a.245131.88

In [7]:
# Find the maximum value in the 'OvertimePay' column of the DataFrame 'sal'.
max_overtime_pay = sal['OvertimePay'].max()
print(max_overtime_pay)

245131.88


6.What is the job title of JOSEPH DRISCOLL ? Note: Use all caps, otherwise you may get an answer that doesn't match up (there is also a lowercase Joseph Driscoll). 
a.24    CAPTAIN, FIRE SUPPRESSION
Name: JobTitle, dtype: object

In [9]:
#  ilter the DataFrame 'sal' to get the 'JobTitle' of the employee with the name 'JOSEPH DRISCOLL'.
name = sal[sal['EmployeeName'] == 'JOSEPH DRISCOLL']['JobTitle']

# Print the 'JobTitle' for the employee with the specified name.
print(name)

24    CAPTAIN, FIRE SUPPRESSION
Name: JobTitle, dtype: object


7. How much does JOSEPH DRISCOLL make (including benefits)?
a.24    270324.91
Name: TotalPayBenefits, dtype: float64

In [10]:
# Filter the DataFrame 'sal' to get the 'TotalPayBenefits' of the employee with the name 'JOSEPH DRISCOLL'.
total_pay_benefits = sal[sal['EmployeeName'] == 'JOSEPH DRISCOLL']['TotalPayBenefits']

# Print the 'TotalPayBenefits' for the employee with the specified name.
print(total_pay_benefits)

24    270324.91
Name: TotalPayBenefits, dtype: float64


8.What is the name of highest paid person (including benefits)?
a.
Id	EmployeeName	JobTitle	BasePay	OvertimePay	OtherPay	Benefits	TotalPay	TotalPayBenefits	Year	Notes	Agency	Status
0	1	NATHANIEL FORD	GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY	167411.18	0.0	400184.25	NaN	567595.43	567595.43	2011	NaN	San Francisco	NaN

In [14]:
# Find the maximum value in the 'TotalPayBenefits' column.
max_total_pay_benefits = sal['TotalPayBenefits'].max()

# Filter the DataFrame 'sal' to get the row(s) where 'TotalPayBenefits' is equal to the maximum value.
max_total_pay_benefits_row = sal[sal['TotalPayBenefits'] == max_total_pay_benefits]

# Print the 'EmployeeName' from the filtered row(s).
print(max_total_pay_benefits_row['EmployeeName'])


0    NATHANIEL FORD
Name: EmployeeName, dtype: object


9.What is the name of lowest paid person (including benefits)? Do you notice something strange about how much he or she is paid?
a.
Id	EmployeeName	JobTitle	BasePay	OvertimePay	OtherPay	Benefits	TotalPay	TotalPayBenefits	Year	Notes	Agency	Status
148653	148654	Joe Lopez	Counselor, Log Cabin Ranch	0.0	0.0	-618.13	0.0	-618.13	-618.13	2014	NaN	San Francisco	NaN

In [15]:
# Find the minimum value in the 'TotalPayBenefits' column.
min_total_pay_benefits = sal['TotalPayBenefits'].min()

# Filter the DataFrame 'sal' to get the row(s) where 'TotalPayBenefits' is equal to the minimum value.
min_total_pay_benefits_row = sal[sal['TotalPayBenefits'] == min_total_pay_benefits]

# Print the details of the row(s) where 'TotalPayBenefits' is at its minimum.
print(min_total_pay_benefits_row)

# Print the name of the lowest-paid person from the filtered row(s).
print(f"\nLowest paid person is: {min_total_pay_benefits_row['EmployeeName'].values[0]}")

# Provide additional information about the observation.
print("The strange thing noticed is basepay, overtimepay is 0, but benefits are in negative.")


            Id EmployeeName                    JobTitle  BasePay  OvertimePay  \
148653  148654    Joe Lopez  Counselor, Log Cabin Ranch      0.0          0.0   

        OtherPay  Benefits  TotalPay  TotalPayBenefits  Year  Notes  \
148653   -618.13       0.0   -618.13           -618.13  2014    NaN   

               Agency  Status  
148653  San Francisco     NaN  

Lowest paid person is: Joe Lopez
The strange thing noticed is basepay, overtimepay is 0, but benefits are in negative.


10.What was the average (mean) BasePay of all employees per year? (2011-2014) ?
a.Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64
11.How many unique job titles are there? 
a.2159

In [16]:
# Group the DataFrame 'sal' by the 'Year' column.
grouped_by_year = sal.groupby('Year')

# Print the data type of the 'Year' column.
print(type(sal.Year))

# Calculate the mean of 'BasePay' for each group (each year).
average_base_pay_by_year = grouped_by_year['BasePay'].mean()

# Print the result.
print(average_base_pay_by_year)



<class 'pandas.core.series.Series'>
Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64


12. What are the top 5 most common jobs? 
a.
Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421
Name: JobTitle, dtype: int64

In [18]:
# Count the occurrences of each unique job title in the 'JobTitle' column of the DataFrame 'sal'.
title_counts = sal['JobTitle'].value_counts()

# Select the top 5 most common job titles based on their occurrence counts.
top_5_titles = title_counts.head(5)

print(top_5_titles)

JobTitle
Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421
Name: count, dtype: int64


13.How many Job Titles were represented by only one person in 2013? (e.g. Job Titles with only one occurence in 2013?) 
a.202

In [20]:
# Filter the DataFrame 'sal' to include only rows where the 'Year' column is equal to 2013.
# Select the 'JobTitle' column for these filtered rows.
filtered_titles_2013 = sal[sal['Year'] == 2013]['JobTitle']

# Count the occurrences of each unique job title in the filtered 'JobTitle' column.
title_counts_2013 = filtered_titles_2013.value_counts()

# Sum the number of titles that have a count of 1, indicating titles with only one occurrence in 2013.
count_single_occurrence_titles_2013 = sum(title_counts_2013 == 1)
print(count_single_occurrence_titles_2013)

202


14.How many people have the word Chief in their job title? (Bonus)
a. 477

In [28]:
chief_count = sal[sal['JobTitle'].str.contains('Chief', case=False, na=False)].shape[0]
print(f"The number of people with the word 'Chief' in their job title is: {chief_count}")


The number of people with the word 'Chief' in their job title is: 627


15. Bonus: Is there a correlation between length of the Job Title string and Salary?
a.
	title_len	TotalPayBenefits
title_len	1.000000	-0.036878
TotalPayBenefits	-0.036878	1.000000

In [30]:
# Creating a new column 'titlelen' in the DataFrame 'df', which contains the length of each 'JobTitle'.
sal['titlelen'] = sal['JobTitle'].apply(len)

# Creating a new DataFrame that contains two columns: 'titlelen' and 'TotalPayBenefits'.
# Calculating the correlation between the two columns.
correlation_result = sal[['titlelen', 'TotalPayBenefits']].corr()
print(correlation_result)

                  titlelen  TotalPayBenefits
titlelen          1.000000         -0.036878
TotalPayBenefits -0.036878          1.000000
